topology/platform/common/ssp.m4

```diff
@@ -43,4 +43,18 @@ define(`SSP_CONFIG_DATA',
 `}'
 )
 
+dnl SSP_BESPOKE_DATA(type, idx, value)
+dnl value is optional
+define(`SSP_BESPOKE_DATA',
+`SectionVendorTuples."$1$2.OUT_tuples_bespoke_w" {'
+`	tokens "sof_dai_tokens"'
+`	tuples."word" {'
+`		SOF_TKN_DAI_BESPOKE_CONFIG'	ifelse($3, `', "0", STR($3))
```

Comment:
The presence of this data implies that there a bespoke config data available so there is no need for a token that says bespoke data available. I'd expect to see tokens in here things that are not generic and specific to SSP only like SSP clock source, M/N dividers, etc 

You could even additionally create an advance mode SSP_BESPOKE_REGS macro that has a token for each register so that register values can be assigned here.

Reply:
The user could then decide if he was using export mode or normal bespoke data, since we are always passing a macro to the DAI macro to expand.

Reply:
@lgirdwood 
This DATA is bind to the SectionWidget for DAI COMP.
We already had some DATA for SectionBE for BE DAI to config SSP clock and other things.
Do you mean we may have other run-time reconfigurable setting to SSP?
That we can use the topology to create kcontrol for some debug use config,

Reply:
Yes, the intention is that we can bind multiple data objects to the widget each containing multiple tuples (but lets start with just 1 data object today). 

he "create loopback kcontrol" flag will be a boolean token in this object and so will any member of the following structure that cant be defined in the standard ALSA topology structure for BE/DAI.

struct sof_ipc_dai_ssp_params {
	uint16_t mode;   // FIXME: do we need this?
	uint16_t mclk_id;

	uint32_t mclk_rate;	/* mclk frequency in Hz */
	uint32_t fsync_rate;	/* fsync frequency in Hz */
	uint32_t bclk_rate;	/* bclk frequency in Hz */

	/* TDM */
	uint32_t tdm_slots;
	uint32_t rx_slots;
	uint32_t tx_slots;

	/* data */
	uint32_t sample_valid_bits;
	uint16_t tdm_slot_width;
	uint16_t reserved2;	/* alignment */

	/* MCLK */
	uint32_t mclk_direction;
	uint32_t mclk_keep_active;
	uint32_t bclk_keep_active;
	uint32_t fs_keep_active;

	//uint32_t quirks; // FIXME: is 32 bits enough ?

	/* private data, e.g. for quirks */
	//uint32_t pdata[10]; // FIXME: would really need ~16 u32
} __attribute__((packed));

e.g. checking the driver I can see we parse for SOF_TKN_INTEL_SSP_MCLK_KEEP_ACTIVE but we don't set this in any private data yet in topology. This token and any others specific to SSP should be added to this data object alongside the "create lbm kcontrol".